code/parachain/frame/composable-traits/src/currency.rs

```diff
@@ -142,13 +163,123 @@ impl<
 {
 }
 
-// hack to imitate type alias until it is in stable
-// named with like implying it is`like` is is necessary to be `AssetId`, but may be not enough (if
-// something is `AssetIdLike` than it is not always asset)
+pub trait AssetIdLike = FullCodec + MaxEncodedLen + Copy + Eq + PartialEq + Debug + TypeInfo;
 
-// FIXME(hussein-aitlahcen): this trait already exists in frame_support, named `AssetId`
-pub trait AssetIdLike:
-	FullCodec + MaxEncodedLen + Copy + Eq + PartialEq + Debug + TypeInfo
-{
+#[derive(
+	RuntimeDebug,
+	Copy,
+	Clone,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Encode,
+	Decode,
+	MaxEncodedLen,
+	TypeInfo,
+)]
+pub struct Rational64 {
+	pub n: u64,
+	pub d: u64,
+}
+
+#[macro_export]
+macro_rules! rational {
+	($n:literal / 0) => {
+		panic!("denominator cannot be zero")
+	};
+	($n:literal / $d: literal) => {
+		Rational64 { n: $n, d: $d }
+	};
+}
+
+macro_rules! impl_const_get {
+	($name:ident, $t:ty) => {
+		#[doc = "Const getter for a basic type."]
+		#[derive(RuntimeDebug)]
+		pub struct $name<const T: $t>;
+		impl<const T: $t> Get<$t> for $name<T> {
+			fn get() -> $t {
+				T
+			}
+		}
+		impl<const T: $t> Get<Option<$t>> for $name<T> {
+			fn get() -> Option<$t> {
+				Some(T)
+			}
+		}
+		impl<const T: $t> TypedGet for $name<T> {
+			type Type = $t;
+			fn get() -> $t {
+				T
+			}
+		}
+	};
+}
```

Comment:
copied from here: https://paritytech.github.io/substrate/master/src/sp_core/lib.rs.html#422

please add docs and link to the original when copying code so maintainers down the line know where it came from